CNTRLPLANE-902:  New metrics for HCP Azure

sdminonne opened the pull request:
**What this PR does / why we need it**:
It adds 2 metrics for Hosted Control Plane (HCP) on Azure.

For Managed HCP it adds metric: `hosted_cluster_managed_azure_info` with the following labels:
- _id
- name
- namespace
- location
- microsoft_subscription_id
- microsoft_resource_group_name
- microsoft_resource_type
- microsoft_resource_id

For not managed HCP it adds metric `hosted_cluster_azure_info` with the following labels:
- _id
- name
- namespace
- location
- microsoft_subscription_id
- microsoft_resource_group_name

The two metrics are mutually exsclusive.
 
Fixes #[CNTRLPLANE-902](https://issues.redhat.com//browse/CNTRLPLANE-902)

**Checklist**
- [x] Subject and description added to both, commit and PR.
- [x] Relevant issues have been referenced.
- [x ] This change includes unit tests.

sdminonne commented:
/retest-required

simonpasquier commented on hypershift-operator/controllers/hostedcluster/metrics/metrics.go:
```diff
@@ -92,6 +92,9 @@ const (
 	ClusterSizeOverrideMetricName = "hypershift_cluster_size_override_instances"
 	clusterSizeOverrideMetricHelp = "Number of HostedClusters with a cluster size override annotation"
+
+	HostedClusterInfoMetricName = "hosted_cluster_info"
```
we might want 2 info metrics
* one `hosted_cluster_info` metric for all generic labels (at least hosted cluster name, namespace, _id and maybe platform type).
* one specific for Azure platform (`hosted_cluster_azure_info`) when platform type = "Azure".

One suggestion I made somewhere else was that we could put everything into "hosted_cluster_info" but it's recommended to have consistent labels per metric. And we might get similar requests in the future for AWS, GCP, ....

sdminonne replied:
thanks @simonpasquier I'll discuss this with my team and let you know.

sdminonne replied:
@devguyio @simonpasquier at the moment we don't have any plan to revamp all metrics and I think I'd say I'm in favor of leaving this metrics as defined in the https://issues.redhat.com/browse/OCPSTRAT-2175

sdminonne commented on hypershift-operator/controllers/hostedcluster/metrics/metrics.go:
```diff
@@ -185,6 +188,14 @@ var (
+
+	hostedClusterInfoDesc = prometheus.NewDesc(
+		HostedClusterInfoMetricName, HostedClusterInfoMetricHelp,
+		append(hclusterLabels, "Microsoft_subscriptionId",
+			"Microsoft_resourceGroupName",
+			"Microsoft_resourceType",
+			"Microsoft_resourceId",
+			"Location"), nil)
```
Note taken, if we'll split the metric we may want to get rid of the prefix

simonpasquier commented on hypershift-operator/controllers/hostedcluster/metrics/metrics.go:
```diff
@@ -185,6 +188,14 @@ var (
+
+	hostedClusterInfoDesc = prometheus.NewDesc(
+		HostedClusterInfoMetricName, HostedClusterInfoMetricHelp,
+		append(hclusterLabels, "Microsoft_subscriptionId",
```
@tony-schndr does it make sense to include other fields of the Azure's HostedCluster spec?
* Cloud
* Location
* VnetID
* SubnetID
* SecurityGroupID
* TenantID

tony-schndr replied:
We need Location, I thought I saw that is already included?  TenantId might also be good to have though it is not required by SRE or Geneva Shoebox.

At the end of the day, everything else is nice to have but SRE has other means to get that information.  The resource ID / region allows us to look everything else up from the RP.

sdminonne replied:
@simonpasquier @tony-schndr thanks for this! It really help? May you this to the https://issues.redhat.com/browse/OCPSTRAT-2175 ? 
Eventually what about changing to `snake_case` as suggested by @devguyio https://github.com/openshift/hypershift/pull/6180#discussion_r2108610241
Ty!

sdminonne commented on hypershift-operator/controllers/hostedcluster/metrics/metrics.go:
```diff
@@ -507,6 +518,25 @@ func (c *hostedClustersMetricsCollector) Collect(ch chan<- prometheus.Metric) {
+				azInfo := hcluster.Spec.Platform.Azure
+				subID := azInfo.SubscriptionID
+				resGroup := azInfo.ResourceGroupName
+				resourceID := fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.RedHatOpenshift/hcpOpenShiftClusters/%s",
```
I'm lost here, I thought all this was for ARO only. Maybe because the title of https://issues.redhat.com/browse/OCPSTRAT-2175 ?
@simonpasquier may you help me to find what would be `resourceID` for HCP in Azure *non*-ARO?

sdminonne commented:
@YuLi517 I'm working on it

sdminonne commented:
@YuLi517 PR (and description) updated. May you double check and eventually let me know?
TY!

YuLi517 commented:
> @YuLi517 PR (and description) updated. May you double check and eventually let me know? TY!

lgtm, thanks @sdminonne

simonpasquier reviewed (commented):
/lgtm

sdminonne commented on hypershift-operator/controllers/hostedcluster/metrics/metrics_test.go:
```diff
@@ -1133,6 +1134,172 @@ func TestReportClusterSizeOverride(t *testing.T) {
+
+	if !found {
+		defer func() {
+			os.Unsetenv("MANAGED_SERVICE")
```
was not aware of this. :TY!

sdminonne commented on hypershift-operator/controllers/hostedcluster/metrics/metrics_test.go:
```diff
@@ -1133,6 +1134,172 @@ func TestReportClusterSizeOverride(t *testing.T) {
 	}
 }
 
+func TestHostedClusterManagedAzureInfo(t *testing.T) {
```
The two metrics are mutually exclusive but I can combine the two tests anyway

sdminonne replied:
Done

bryan-cox commented on hypershift-operator/controllers/hostedcluster/metrics/metrics.go:
```diff
@@ -507,6 +531,35 @@ func (c *hostedClustersMetricsCollector) Collect(ch chan<- prometheus.Metric) {
+				subID := azInfo.SubscriptionID
+				resGroup := azInfo.ResourceGroupName
+				if azureutil.IsAroHCP() {
+					resourceID := fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.RedHatOpenshift/hcpOpenShiftClusters/%s",
```
I think adding a comment to the permalink would work for this comment 👍🏻

sdminonne commented:
@bryan-cox 

- modified the commtis comment (PTAL), 
- added permalink to  ARO `resource_id` resource
- get rid of `os.Setenv` replaced with `t.Setenv`
- merged the 2 tests

PTAL

bryan-cox commented:
/approve

devguyio commented:
/lgtm
